src/wbase/UserQueryInfo.h

```diff
@@ -0,0 +1,123 @@
+// -*- LSST-C++ -*-
+
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_WBASE_USERQUERYINFO_H
+#define LSST_QSERV_WBASE_USERQUERYINFO_H
+
+// System headers
+#include <map>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+// Qserv headers
+#include "global/intTypes.h"
+
+
+// This header declarations
+namespace lsst::qserv::wbase {
+
+/// &&& doc
+class UserQueryInfo {
```

Comment:
There are two collection objects that are important (both have changed to make them safe):

The first is the static map, `_uqMap`,  which contains weak pointers to all UserQueryInfo instances and is ordered by `QueryId`. `QueryId` is the same for all `Task` and `Job` objects in the UserQuery. `_uqMapMtx` is used to protect `_uqMap` from race conditions. Before, it used normal shared pointers and wouldn't delete the map entry when the UserQueryInfo was no longer needed. That should be fixed now.

The second is the vector of template strings, `_templates`. The index into the vector is not a pointer, it's just an unsigned integer index pointing to a std::string. Even if the vector is moved due to needing more memory, the index to a string should still be valid. New strings are always added to the end, so the previous indexes will not change. `_uqMtx` is now being used to prevent race conditions.

`_templates` is kind of annoying as it seems like it would be really nice to use the template string as the key so it can be found quickly when searching. However, using the template as the key completely defeats the purpose of saving space, and as you pointed out, putting them in a `set` or `map` could cause them to move and invalidate the indexes. Luckily, there are only a small number of templates  for any UserQuery (I think 2 max, but I'm not certain), so searching a `vector` for a string match is not much worse than a map or set.